src/imgui/imgui_raii_wrapper.h

```diff
@@ -0,0 +1,162 @@
+};
+
+class TabBar {
+public:
+  [[nodiscard]] explicit TabBar(std::string const& id) : selected{ImGui::BeginTabBar(id.c_str())} {}
+  TabBar(TabBar const&) = delete;
+  TabBar& operator=(TabBar const&) = delete;
+  TabBar(TabBar&&) = delete;
+  TabBar& operator=(TabBar&&) = delete;
+  ~TabBar() {
+    if (selected) {
+      ImGui::EndTabBar();
+    }
+  }
+  explicit operator bool() & {
+    return selected;
+  }
+private:
+  bool selected{};
+};
+
+class TabItem {
+public:
+  [[nodiscard]] explicit TabItem(std::string const& id) : selected{ImGui::BeginTabItem(id.c_str())} {}
+  TabItem(TabItem const&) = delete;
+  TabItem& operator=(TabItem const&) = delete;
+  TabItem(TabItem&&) = delete;
+  TabItem& operator=(TabItem&&) = delete;
+  ~TabItem() {
+    if (selected) {
+      ImGui::EndTabItem();
+    }
+  }
+  explicit operator bool() & {
+    return selected;
+  }
+private:
+  bool selected{};
+};
+
+class Text {
+public:
+  Text(char const* text_) : text{text_} {}
+  Text(std::string text_) : text{std::move(text_)} {}
+  Text(char const* text_, ImVec4 color_) : text{text_}, color{color_} {}
+  Text(std::string text_, ImVec4 color_) : text{std::move(text_)}, color{color_} {}
+  std::string get_text() const {
+    return text;
+  }
+  std::optional<ImVec4> get_color() const {
+    return color;
+  }
+  void write() const {
+    if (color) {
+      ImGui::TextColored(color.value(), "%s", text.c_str());
+    } else {
+      ImGui::Text("%s", text.c_str());
+    }
+  }
+private:
+  std::string text{};
+  std::optional<ImVec4> color{};
+};
+
+class Table {
+public:
+  using Row = std::vector<Text>;
+  class Builder {
+  public:
+    Builder& with_id(std::string name) {
+      id = std::move(name);
+      return *this;
+    }
+    Builder& with_number_of_columns(std::size_t n) {
+      number_of_columns = n;
+      return *this;
+    }
+    Builder& with_headers(Row row) {
+      headers = std::move(row);
+      return *this;
+    }
+    Builder& add_row(Row row) {
+      values.push_back(std::move(row));
+      return *this;
+    }
+    Table build() {
+      return Table{
+        id,
+        static_cast<int>(number_of_columns),
+        headers,
+        values
+      };
+    }
+  private:
+    std::string id{};
+    std::size_t number_of_columns{};
+    Row headers{};
+    std::vector<Row> values{};
+  };
+
+  explicit Table(std::string const& id, int number_of_columns, Row const& headers, std::vector<Row> const& values) : selected{ImGui::BeginTable(id.c_str(), number_of_columns)} {
+    create_headers(headers);
+    for (auto const& row : values) {
+      create_row(row);
+    }
+  }
+  Table(Table const&) = delete;
+  Table& operator=(Table const&) = delete;
+  Table(Table&&) = delete;
+  Table& operator=(Table&&) = delete;
+  ~Table() {
+    ImGui::EndTable();
+  }
+  explicit operator bool() & {
+    return selected;
+  }
+private:
+  static void create_headers(Row const& labels) {
+    for (auto const& label : labels) {
```

Comment:
Is this an automatic constructor?

Reply:
No. There's no constructor being called here. And I don't know what an automatic constructor is supposed to be. As you know, in C++ you have to specify the type of every variable. `auto` means the compiler will try to deduce the type for you. `const` is a promise that label won't be changed.

Reply:
I guess by now you already know what `const` is.
`auto` means that the type of the var being declared will be automatically deduced. it comes handy for situations like this or when you have complicated and verbose types, like `iterators`